rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,191 @@
+# RFC 14742 - Support for Duplicate and Bare Tags on Metrics
+
+Vector's current metric model supports only a single value for each tag. The `datadog_agent` source,
+however, may send metrics that contain multiple values for a given tag name, or bare tags with no
+value. Vectors tag support should be enhanced to handle this data.
+
+## Context
+
+- [Support duplicate tag and bare tags on metrics](https://github.com/vectordotdev/vector/issues/14742)
+- [DataDog Source -> Filter Transform -> DataDog Sink dropping kube_service tag](https://github.com/vectordotdev/vector/issues/14707)
+- [Invalid custom_tags metrics from sinks datadog_metric](https://github.com/vectordotdev/vector/issues/14239)
+
+## Scope
+
+### In scope
+
+- Changes to the existing metric tag internal and external representation to support multiple values
+  and bare tags.
+
+### Out of scope
+
+- Changes to any other aspect of metric or event representations.
+
+## Pain
+
+When ingesting data from a Datadog agent, the metric tags provided by that source may contain data
+that Vector's data model cannot represent. This is causing problems for prospective users that want
+to add Vector to their observability pipelines but cannot do so because it would cause data loss.
+
+## Proposal
+
+### User Experience
+
+This change should be effectively invisible to users of Vector, except for incoming data with
+multiple tags with the same name, which will now have all the tags reproduced when it reaches sinks.
+
+### Implementation
+
+#### Internal Representation
+
+Simply put, the tags representation will change from an alias into a newtype wrapper using an
+`indexmap` set to store the tag values. This newtype will hide the implementation details of the
+underlying storage from the callers. It will also add separate methods for inserting a new tag and
+replacing a tag set with a single value. The callers of the existing `insert` function will need to
+be audited to determine which use is intended at each call site.
+
+```rust
+type TagValue = Option<String>;
+
+struct MetricTags(BTreeMap<String, indexmap::IndexSet<TagValue>>);
+
+impl MetricTags {
+    // Insert returns the value unchanged if the exact (tag,value) pair already exists,
+    // otherwise it inserts a new value for the named tag.
+    fn insert(&mut self, name: String, value: TagValue) -> Option<TagValue>;
+
+    // Replace returns all the existing values when overwriting a tag.
+    fn replace(&mut self, name: String, value: Option<TagValue>) -> Option<IndexSet<TagValue>>;
+}
```

Comment:
I was surprised I didn't see methods here to delete tags as well as fetch them. Are these unneeded?